Streaming.cpp

```diff
@@ -756,139 +750,141 @@ int SoapySidekiq::deactivateStream(SoapySDR::Stream *stream, const int flags,
-    }
-    else
-    {
-        timeNs = convert_timestamp_to_nanos(block_ptr->sys_timestamp, sys_freq);
-    }
-
-    flags = SOAPY_SDR_HAS_TIME;
-
-    uint32_t block_num = 0;
-    uint32_t num_blocks = numElems / rx_payload_size_in_words;
+    if (stream != RX_STREAM) return SOAPY_SDR_NOT_SUPPORTED;
+    if (rx_receive_operation_exited_due_to_error) return SOAPY_SDR_STREAM_ERROR;
+
+    size_t samples_done = 0;
+    bool timestamp_set = false;
+    long waitTime = (timeoutUs == 0) ? SLEEP_1SEC : timeoutUs;
+
+    // Output pointer (void*, could be int16_t* or float*)
+    void *output_buf = buffs[0];
+
+    while (samples_done < numElems) {
+        // Consume any FIFO buffer leftovers first
+        size_t fifo_left = (rx_fifo_buffer.size() / 2) - rx_fifo_offset;
+        if (fifo_left > 0) {
+            size_t to_copy = std::min(fifo_left, numElems - samples_done);
+
+            if (rxUseShort) {
+                int16_t *out_ptr = reinterpret_cast<int16_t *>(output_buf);
+                memcpy(
+                    out_ptr + samples_done * 2,
+                    rx_fifo_buffer.data() + rx_fifo_offset * 2,
+                    to_copy * 2 * sizeof(int16_t));
+            } else {
+                float *out_ptr = reinterpret_cast<float *>(output_buf);
+                for (size_t i = 0; i < to_copy; ++i) {
+                    out_ptr[(samples_done + i) * 2]     = float(rx_fifo_buffer[(rx_fifo_offset + i) * 2])     / float(this->maxValue);
+                    out_ptr[(samples_done + i) * 2 + 1] = float(rx_fifo_buffer[(rx_fifo_offset + i) * 2 + 1]) / float(this->maxValue);
+                }
+            }
+            rx_fifo_offset += to_copy;
+            samples_done += to_copy;
 
-    while (block_num < num_blocks)
-    {
-        // copy in the amount of data we have in the ring block
-        if (rxUseShort == true)
-        {
-            // CS16
-            memcpy(buff_ptr, ringbuffer_ptr, rx_payload_size_in_bytes);
+            if (rx_fifo_offset * 2 >= rx_fifo_buffer.size()) {
+                rx_fifo_buffer.clear();
+                rx_fifo_offset = 0;
+            }
+            continue;
         }
-        else
-        {
-            // CF32 float
-            float *  dbuff_ptr = (float *)buff_ptr;
-            int16_t *source = (int16_t *)ringbuffer_ptr;
-            int short_ctr = 0;
-            uint32_t i;
 
-            for (i = 0; i < rx_payload_size_in_words; i++)
-            {
-                *dbuff_ptr++ = (float)source[short_ctr + 1] / this->maxValue;
-                *dbuff_ptr++ = (float)source[short_ctr] / this->maxValue;
-                short_ctr += 2;
-            }
+        // Wait for a new block in the ring buffer
+        while ((rxReadIndex == rxWriteIndex) && (waitTime > 0)) {
+            usleep(DEFAULT_SLEEP_US);
+            waitTime -= DEFAULT_SLEEP_US;
+        }
+        if (waitTime <= 0) {
+            return (samples_done > 0) ? samples_done : SOAPY_SDR_TIMEOUT;
```

Comment:
So if we timeout should we not tell the calling user?  In this case you are returning the amount of samples received.